[DOC] tempo-distributed helm chart doc (#2127)
grafana/tempo: docs/sources/helm-charts/tempo-distributed/get-started-helm-charts/_index.md

---
description: Learn how to get started with Grafana Tempo using the Helm chart.
menuTitle: Get started
title: Get started with Grafana Tempo using the Helm chart
weight: 20
keywords:
  - Helm chart
  - Kubernetes
  - Grafana Tempo
---

# Get started with Grafana Tempo using the Helm chart

The Grafana Tempo Helm chart allows you to configure, install, and upgrade Grafana Tempo within a Kubernetes cluster. Using this procedure, you will:

> - Create a custom name-space within your Kubernetes cluster

Reviewer: I think just `namespace`, as it is referred to in kubernetes.

Author: Done

- Configure Google Cloud Platform storage for traces
- Install Tempo using Helm

To learn more about Helm, read the [Helm documentation](https://helm.sh/).

## Before you begin

These instructions are common across any flavor of Kubernetes. They also assume that you know how to install, configure, and operate a Kubernetes cluster.
It also assumes that you have an understanding of what the `kubectl` command does.

>**CAUTION**: This procedure is primarily aimed at local or development setups.

### Hardware requirements

- A single Kubernetes node with a minimum of 4 cores and 16GiB RAM

### Software requirements

- Kubernetes 1.20 or later (see [Kubernetes installation documentation](https://kubernetes.io/docs/setup/))
- The `kubectl` command for your version of Kubernetes
- Helm 3 or later (see [Helm installation documentation](https://helm.sh/docs/intro/install/))

Verify that you have:
- Access to the Kubernetes cluster
- Persistent storage is enabled in the Kubernetes cluster, which has a default storage class setup. You can [change the default StorageClass](https://kubernetes.io/docs/tasks/administer-cluster/change-default-storage-class/).

> - Access to a Google Cloud Platform storage bucket (see [instructions](https://cloud.google.com/storage/docs/creating-buckets))

Reviewer: This requirement breaks the local caution above.  Also, we probably want to say something to users who are in Azure or AWS, and not in GCP.

Author: I think we've addressed this comment by reorganizing the content for this procedure.

1. Set up a Helm repository using the following commands:

   ```bash
   helm repo add grafana https://grafana.github.io/helm-charts
   helm repo update
   ```

   >**NOTE**: The Helm chart at https://grafana.github.io/helm-charts is a publication of the source code at grafana/tempo.

### Set Google Cloud Platform as the storage option

Before you run the Helm chart, you need to configure where trace data will be stored. These steps create a YAML file with the storage configuration.

You need the `secret_key` from your service provider used as part of the user or account authentication. In this case, it refers to S3’s secret key. Google Cloud Storage has a similar option.

1. Create a YAML file of Helm values called `custom.yaml`.

1. Add the following configuration to the file to set Google Cloud Platform as the trace storage.
   - Change `temorockstracing` to your `secret_key`.
   - Change `bucket` to match the bucket name in GCP.
   - Change `endpoint` to the correct one for your set up.

   ```yaml
   ---
   minio:
     enabled: false
   storage:
     trace:
       backend: s3

>        s3:

Reviewer: We'll want to use a GCP example as per the above instructions.

         access_key: 'tempo-test'
         secret_key: 'temporockstracing'
         bucket: 'tempo-test'

>          endpoint: 'stodev0.znet:9000'

Reviewer: Endpoint will only be necessary I think in the case of minio, or a custom s3 endpoint.

If you are using the PodSecurityPolicy admission controller, then it is not possible to deploy the mimir-distributed chart with MinIO. Refer to Run Grafana Mimir in production using the Helm chart for instructions on setting up an external object storage and disable the built-in MinIO deployment with minio.enabled: false in the Helm values file.

If you are using the Pod Security admission controller, then MinIO and the tempo-distributed chart can successfully deploy under the baseline pod security level.
-->

## Create a custom namespace and add the Helm repository

Using a custom namespace solves problems later on because you do not have to overwrite the default namespace.

> 1. Create a unique Kubernetes namespace, for example tempo-test:

Author: ```suggestion
1. Create a unique Kubernetes namespace, for example `tempo-test`:
```

```bash
kubectl create namespace tempo-test
```

   For more details, see the Kubernetes documentation about [Creating a new namespace](https://kubernetes.io/docs/tasks/administer-cluster/namespaces/#creating-a-new-namespace).

1. Set up a Helm repository using the following commands:

   ```bash
   helm repo add grafana https://grafana.github.io/helm-charts
   helm repo update
   ```

   >**NOTE**: The Helm chart at https://grafana.github.io/helm-charts is a publication of the source code at grafana/tempo.

## Set Helm chart values

> Your Helm chart values are set in the `custom.yaml` file. The following example `custom.yaml` file sets the storage and traces options, enables the gateway, and sets the cluster to main. The `traces` configure the distributor's receiver protocols.

Author: @zalegrala I referenced the GET Helm chart doc and we point people to this file: https://github.com/grafana/helm-charts/blob/main/charts/tempo-distributed/values.yaml 

Should we be using that values.yaml file instead of creating a custom.yaml file? The values.yaml file has the section for MinIO already there. We'd just have to tell people to change the value to enabled.

> Your Helm chart values are set in the `custom.yaml` file. The following example `custom.yaml` file sets the storage and traces options, enables the gateway, and sets the cluster to main. The `traces` configure the distributor's receiver protocols.

Author: ```suggestion
All `tempo-distributed` Helm chart options are listed in the [`values.yaml`](https://github.com/grafana/helm-charts/blob/main/charts/tempo-distributed/values.yaml) file. 

For this procedure, we're using the `custom.yaml` file to set the values for our Helm chart. 
The `custom.yaml` file acts as an override to items set in the `values.yaml` file. 
The following example `custom.yaml` file sets the storage and traces options, enables the gateway, and sets the cluster to `main`. 
The `traces` configure the distributor's receiver protocols.
```

```yaml
---
storage:
  trace:
    backend: s3
    s3:
      access_key: 'grafana-tempo'
      secret_key: 'supersecret'
      bucket: 'tempo-traces'
      endpoint: 'tempo-minio:9000'
      insecure: true

>  traces:

Author: ```suggestion
minio:
  enabled: true
  mode: standalone
  rootUser: minio
  rootPassword: minio123
 traces:
```

Author: @zalegrala Is this the correct configuration block for MinIo? Do we also need anything else?

  otlp:
    grpc:
      enabled: true
    http:
      enabled: true
  zipkin:
    enabled: false
  jaeger:
    thriftHttp:
      enabled: false
  opencensus:
    enabled: false
```

Next, we will:

> 1. Create a `custom.yaml` file

Author: ```suggestion
1. Save a local copy of the `values.yaml` file. 
```

2. Set your storage values, the example above points to the MinIO instance configured by the chart
3. Set your traces values to configure the receivers on the Tempo distributor

### Create a `custom.yaml` file

> 1. Create a YAML file of Helm values called `custom.yaml`.
> 1. Copy the example above and paste it into the custom.yaml file.

Author: ```suggestion
1. Create a YAML file of Helm values called `values.yaml`.
1. Copy the [`values.yaml`](https://github.com/grafana/helm-charts/blob/main/charts/tempo-distributed/values.yaml) contents and paste it into your local `values.yaml` file.
```

1. Save the file.

### Set your storage option

Before you run the Helm chart, you need to configure where trace data will be stored.

The `storage` block defined in the `values.yaml` file is used to configure the storage that Tempo uses for trace storage.

The procedure below configures MinIO as the local storage option managed by the Helm chart.
However, you can use a other storage provides. Refer to the Optional storage section below.

> 1. Update the configuration options in `custom.yaml` for your configuration.

Author: ```suggestion
1. Update the configuration options in `values.yaml` for your configuration.
```

> 1. Update the configuration options in `custom.yaml` for your configuration.

Author: ```suggestion
1. Update the configuration options in `values.yaml` for your configuration.
```

> 1. Update the configuration options in `custom.yaml` for your configuration.
> 
>    ```yaml
>    ---
>    storage:
>      trace:
>        backend: s3
>        s3:
>          access_key: 'grafana-tempo'
>          secret_key: 'supersecret'
>          bucket: 'tempo-traces'
>          endpoint: 'tempo-minio:9000'
>          insecure: true
>    ```

Author: ```suggestion
1. Open your `values.yaml` file in an editor. 
1. Locate the `minio` section. 
1. Enable MinIO usage and change the default username and password:
  minio:
    enabled: true
    mode: standalone
    rootUser: minio
    rootPassword: minio123

1. Update the configuration options in `custom.yaml` for your configuration.

   ```yaml
   ---
   storage:
     trace:
       backend: s3
       s3:
         access_key: 'grafana-tempo'
         secret_key: 'supersecret'
         bucket: 'tempo-traces'
         endpoint: 'tempo-minio:9000'
         insecure: true
   ```
```

> 

Author: ```suggestion

1. Optional: Locate the MinIO section and change the the username and password to something you wish to use. 
```yaml
  minio:
    enabled: true
    mode: standalone
    rootUser: minio
    rootPassword: minio123
 ```
```

> ### Optional: Other storage option

Author: ```suggestion
### Optional: Other storage options
```

- [Azure Blob Storage configuration documentation](https://grafana.com/docs/tempo/latest/configuration/azure/)

- [Google Cloud Storage configuration documentation](https://grafana.com/docs/tempo/latest/configuration/gcs/)

<!--- GET example to document later
```yaml
minio:
  enabled: true
storage:
  trace:
    backend: s3
    s3:
      access_key: 'grafana-tempo'
      secret_key: 'supersecret'
      bucket: 'enterprise-traces'
      endpoint: 'tempo-minio:9000'
      insecure: true
  admin:
    backend: s3
    s3:
      access_key_id: 'grafana-tempo'
      secret_access_key: 'supersecret'
      bucket_name: 'enterprise-traces-admin'
      endpoint: 'tempo-minio:9000'
      insecure: true
```
-->

### Set traces receivers

Tempo can be configured to receive data from OTLP, Jaegar, Zipkin, Kafka, and OpenCensus.
 The following example enables OTLP on the distributor. For other options, refer to the [distributor documentation](https://grafana.com/docs/tempo/latest/configuration/#distributor)

The example used in this procedure has OTLP enabled.

> 

Author: ```suggestion

1. Open your `values.yaml` file. 
1. Locate the `traces:` section. 
1. Enable any protocols you wish to use. In the example below, only OTLP is enabled. 
```

> Enable any other protocols based on your requirements.

Author: ```suggestion

```

 ```yaml
 traces:
  otlp:
    grpc:
      enabled: true
    http:
      enabled: true
  zipkin:
    enabled: false
  jaeger:
    thriftHttp:
      enabled: false
  opencensus:
    enabled: false
```

> 

Author: 1. Save changes.

### Optional: Add custom configurations

> You can use a YAML file, like `custom.yaml`, to store custom configuration options that override the defaults present in the Helm chart.
> The [tempo-distributed Helm chart's README](https://github.com/grafana/helm-charts/blob/main/charts/tempo-distributed/README.md) contains a list of available options.

Author: ```suggestion
You can change other sections in the `values.yaml` file to customize the Helm chart values. Using the `values.yaml` file, as described in this procedure, ensures that you can modify any available settings. 

The [tempo-distributed Helm chart's README](https://github.com/grafana/helm-charts/blob/main/charts/tempo-distributed/README.md) contains a list of available options.
```

To see all of the configurable parameters for the `tempo-distributed` Helm chart, use the following command:

```bash
helm show values grafana/tempo-distributed
```

> The configuration sections are added to the `custom.yaml` file. This file is included when you install or upgrade the Helm chart.

Author: ```suggestion
The configuration sections are added or udpated to the `values.yaml` file. This file is included when you install or upgrade the Helm chart.
```

#### Optional: Configure an ingress

An ingress lets you externally access a Kubernetes cluster.
Replace `<ingress-host>` with a suitable hostname that DNS can resolve to the external IP address of the Kubernetes cluster.
For more information, see [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/).

>**NOTE**: On Linux systems, and if it is not possible for you set up local DNS resolution, you can use the `--add-host=<ingress-host>:<kubernetes-cluster-external-address>` command-line flag to define the `<ingress-host>` local address for the docker commands in the examples that follow.

> 1. Open your `custom.yaml` or create a YAML file of Helm values called `custom.yaml`.

Author: ```suggestion
1. Open your `values.yaml`.
```

> 1. Add the following configuration to the file:

Author: ```suggestion
1. Locate the `nginx:` section in the chart. 
1. Make the following configuration changes:
```

   ```
   nginx:
     ingress:
       enabled: true
       ingressClassName: nginx
       hosts:
         - host: <ingress-host>
           paths:
             - path: /
               pathType: Prefix
       tls:
         # empty, disabled.
   ```

> 

Author: ```suggestion
1. Save the changes. 
```

## Install Grafana Tempo using the Helm chart

> Use the following command to install Tempo using the configuration options you’ve specified in the `custom.yaml` file:

Author: ```suggestion
Use the following command to install Tempo using the configuration options you’ve specified in the `values.yaml` file:
```

```bash

> helm -n tempo-test install tempo grafana/tempo-distributed -f custom.yaml

Author: ```suggestion
helm -n tempo-test install tempo grafana/tempo-distributed -f values.yaml
```

```

>**NOTE**: The output of the command contains the write and read URLs necessary for the following steps.

If the installation is successful, the output will be similar to this:

```yaml

> >  helm -n tempo-test install tempo grafana/tempo-distributed -f custom.yaml

Author: ```suggestion
>  helm -n tempo-test install tempo grafana/tempo-distributed -f values.yaml
```

---
description: Learn how to get started with Grafana Tempo using the Helm chart.
menuTitle: Get started
title: Get started with Grafana Tempo using the Helm chart
weight: 20
keywords:
  - Helm chart
  - Kubernetes
  - Grafana Tempo
---

# Get started with Grafana Tempo using the Helm chart

The Grafana Tempo Helm chart allows you to configure, install, and upgrade Grafana Tempo within a Kubernetes cluster. Using this procedure, you will:

> - Create a custom name-space within your Kubernetes cluster

Author: ```suggestion
- Create a custom namespace within your Kubernetes cluster
```

W0210 15:02:09.930379    8613 warnings.go:70] spec.template.spec.topologySpreadConstraints[0].topologyKey: failure-domain.beta.kubernetes.io/zone is deprecated since v1.17; use "topology.kubernetes.io/zone" instead
NAME: tempo
LAST DEPLOYED: Fri Feb 10 15:02:08 2023
NAMESPACE: tempo-test
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
***********************************************************************
 Welcome to Grafana Tempo
 Chart version: 1.0.1
 Tempo version: 2.0.0
***********************************************************************

Installed components:
* ingester
* distributor
* querier
* query-frontend
* compactor
* memcached
```

>**NOTE**: If you update your `values.yaml` or `custom.yaml`, run the same helm install command and replace `install` with `upgrade`.

Check the statuses of the Tempo pods:

```bash
kubectl -n tempo-test get pods
```

The results look similar to this:

```
NAME                                    READY   STATUS    RESTARTS   AGE
tempo-compactor-86cd974cf-8qrk2         1/1     Running   0          22h
tempo-distributor-bbf4889db-v8l8r       1/1     Running   0          22h

> tempo-gateway-75f468bd46-fbzj4          1/1     Running   0          22h

Reviewer: Delete this line to be consistent with the values we provided.  This was a carry-over from a previous `helm install` on my part I think.

Author: Done.

> tempo-gateway-75f468bd46-fbzj4          1/1     Running   0          22h

Author: ```suggestion
```

tempo-ingester-0                        1/1     Running   0          22h
tempo-ingester-1                        1/1     Running   0          22h
tempo-ingester-2                        1/1     Running   0          22h
tempo-memcached-0                       1/1     Running   0          8d
tempo-minio-6c4b66cb77-sgm8z            1/1     Running   0          26h
tempo-querier-777c8dcf54-fqz45          1/1     Running   0          22h
tempo-query-frontend-7f7f686d55-xsnq5   1/1     Running   0          22h
```

Wait until all of the pods have a status of Running or Completed, which might take a few minutes.

## Next step
The next step is to test your Tempo installation by sending trace data to Grafana. You can use the [Set up a test application for a Tempo cluster](https://grafana.com/docs/tempo/latest/setup/set-up-test-app/) document for step-by-step instructions.

If you already have Grafana available, you can add a Tempo data source using the URL fitting to your environment. For example:

> `http://tempo-query-frontend.trace-test.svc.cluster.znet:3100`

Reviewer: change `cluster.znet` to `cluster.local` :)

Author: Done. 
```suggestion
`http://tempo-query-frontend.trace-test.svc.cluster.local:3100`
```

---
description: Learn how to get started with Grafana Tempo using the Helm chart.
menuTitle: Get started
title: Get started with Grafana Tempo using the Helm chart

> weight: 20

Author: ```suggestion
draft: true
weight: 20
```
